WIP: Cadastro básico, visando apenas apenas validar o funcionamento

arthurlima96 opened the pull request:
Cadastro do modelo usuário, criar um entregável de ponta a ponta

arthurlima96 committed:
Cadastro básico, visando apenas apenas validar o funcionamento

asouza commented on nosso-calendario/src/main/java/com/br/devefc/nossocalendario/model/User.java:
```diff
@@ -0,0 +1,55 @@
+	public User() {
+	}
+	
+	public User(String login, String senha, LocalDateTime acesso_cadastro) {
```
*Deixamos pistas que facilitem o uso do código onde não conseguimos resolver com compilação.*

O login é obrigatório e formato de email e a senha é obrigatória e precisa ter no máximo 6 caracteres. O seu construtor, neste momento não valida(que é chato mesmo) nem indica para quem vai invocar que essas informações tem características que precisam ser respeitadas. Você pode usar as annotations da BeanValidation direto no construtor.

asouza replied:
Ainda aqui: se você recebe o dia de acesso de cadastro como argumento você autoriza que alguem seja cadastrado num instante que não representa o atual, o que fere a regra de negócio da funcionalidade. O instante de cadastro tem que ser o instante atual. Sugiro não receber isso como argumento.

asouza replied:
Mais uma coisa: Como você sabe que a senha ali está vindo com o hash aplicado? Lembra que não queremos salvá-la de forma limpa no banco.

asouza commented on nosso-calendario/src/main/java/com/br/devefc/nossocalendario/model/User.java:
```diff
@@ -0,0 +1,55 @@
+	@GeneratedValue(strategy=GenerationType.AUTO)
+	private Long id;
+
+	private String login;
```
*Protegemos as bordas do sistema como se não houvesse amanhã.*

Alguma hora esse objeto vai ser serializado e salvo no banco de dados pelo Hibernate. Minha sugestão é colocar as annotations da BeanValidation também nos atributos, já que o Hibernate vai validar esses dados antes de mandá-los para o banco.  Dessa forma você adiciona mais uma camada de validação e garante que sua aplicação não vai trabalhar com dados inválidos.

asouza commented on nosso-calendario/src/main/java/com/br/devefc/nossocalendario/model/User.java:
```diff
@@ -0,0 +1,55 @@
+		this.acesso_cadastro = acesso_cadastro;
+	}
+
+	public Long getId() {
```
*A complicação do nosso é código é proporcional a complicação da nossa feature. Quanto mais simples, melhor.*

Se tem método público na sua classe é porque alguém, provavelmente vai chamar. O que não é o caso neste momento do seu código. Minha sugestão aqui é apagar e esperar a necessidade aparecer :).

arthurlima96 replied:
> _A complicação do nosso é código é proporcional a complicação da nossa feature. Quanto mais simples, melhor._
> 
> Se tem método público na sua classe é porque alguém, provavelmente vai chamar. O que não é o caso neste momento do seu código. Minha sugestão aqui é apagar e esperar a necessidade aparecer :).

Optei por deixar disponível somente o método **getId()**, mais por causa da visualização no swagger e caso for escrever algum teste eu validar o retorno se foi salvo ou não por meio do id retornado

asouza commented on nosso-calendario/src/main/java/com/br/devefc/nossocalendario/rest/UserController.java:
```diff
@@ -0,0 +1,26 @@
+	private UserService userService;
+
+	@PostMapping(path = "/salvar")
+	public ResponseEntity<User> save(@RequestBody User user) {
```
*Quanto mais externa a borda, mais proteção temos.*
*Usamos tudo que conhecemos que está pronto. Só fazemos código do zero se for estritamente necessário.*

Aqui você está na borda mais externa da aplicação e você **precisa** garantir que não vai entrar nada inválido na sua aplicação. O Spring já valida a entrada dos parâmetros automaticamente para você, através da Bean Validation.  Basta marcar o parâmetro com @Valid. Neste momento você está aceitando usuário com senha vazia e login vazio.

asouza replied:
*Separamos as bordas externas do sistema do seu núcleo. Não ligamos parâmetros de requisição externa com objetos de domínio diretamente.*

Aqui vai entrar como uma otimização prematura, mas que na minha opinião tende a ser benéfica. Neste exato momento a sua classe de domínio já possui métodos que ninguém da aplicação usa, só o Spring. Os setters e getters(esses ninguém usa mesmo por enquanto). Em algum momento o formulário fica diferente do modelo em si e aí, neste tipo de caso, você fica tentado a mudar o domínio para atender o formulário. Só que esta mesma classe de dominio já está sendo usada em outros lugares do projeto e pronto, a confusão começa :). 

Dá uma olhada nesse vídeo aqui => https://www.youtube.com/watch?v=lq5xpz79i5Q&list=PLVHlvMRWE0Y70QZLWvQvDxxQtFeOEHFkY&index=5&t=726s da playlist do meu canal, acho que pode ajudar.

Este post também ajuda => https://spring.io/guides/gs/validating-form-input/ Só não precisa do BindingResult neste momento. 

A ideia é que você o método do controller possa ser invocado independente do seu domínio, pq os parâmetros só tem conexão realmente com a web.

arthurlima96 replied:
Refatoração implementada, dessa forma eu consegui separar bem os domínios,  e consegui ocultar campos que não que não faziam sentido de serem exibidos na API

asouza commented on nosso-calendario/src/main/java/com/br/devefc/nossocalendario/rest/UserController.java:
```diff
@@ -0,0 +1,26 @@
+public class UserController {
+
+	@Autowired
+	private UserService userService;
```
*A complicação do nosso é código é proporcional a complicação da nossa feature. Quanto mais simples, melhor.*

O seu service, chama o método que salva no repositório. Ele não faz nada além disso. Aqui o código fica com mais um passo para que a pessoa seja capaz de entender o que está acontecendo realmente. Minha sugestão é usar o Repository direto aqui... Faça o seguinte exercício: se você não tivesse lido ou usado esse padrão antes, qual é o argumento realmente lógico que existe para usar o Service aqui? O que ele facilita?

arthurlima96 replied:
Realmente, por conta das convenções, costumamos ficar habituados a dividir o sistema por meio dessas camandas

asouza commented on nosso-calendario/src/main/java/com/br/devefc/nossocalendario/service/UserService.java:
```diff
@@ -0,0 +1,18 @@
+import com.br.devefc.nossocalendario.repository.UserRepository;
+
+@Component
+public class UserService {
```
Sugiro apagar :)

Commit:
Adição da validação dos campos de email e senha com Bean Validation

Commit:
Adição do Bean validation no construtor e remoção de modificação acesso

asouza commented on nosso-calendario/src/main/java/com/br/devefc/nossocalendario/model/User.java:
```diff
@@ -15,17 +20,31 @@
 	@GeneratedValue(strategy=GenerationType.AUTO)
 	private Long id;
 
+	@NotBlank(message = "Login Obrigatorio")
```
Lembra que a mensagem, normalmente, pode ficar apenas nas validações de borda mais externa, que é onde nos recuperamos do problema de validação e retornarmos algo para o usuário. Nesse momento aqui, se chegar algo inválido, é bug mesmo hehe.

asouza reviewed (changes requested):
Acho que não era para eu revisar ainda, né? foi mal.

arthurlima96 commented:
> Acho que não era para eu revisar ainda, né? foi mal.

Tranquilo kk, pode revisar sim, pq tenho o costume de fazer e ir logo commitando

Commit:
Remoção do Bean Validation do Construtor

arthurlima96 committed:
Aplicação do desafio da senha com o hash e melhorias sugeridas no modelo

arthurlima96 committed:
Remoçao da camada de serviço

arthurlima96 committed:
Refatoração com criação do UserForm para uso no Controller

asouza commented on nosso-calendario/src/main/java/com/br/devefc/nossocalendario/model/User.java:
```diff
@@ -12,59 +12,62 @@
 	private Long id;
 
 	@NotBlank(message = "Login Obrigatorio")
 	@NotEmpty(message = "Login não pode ser vazio")
```
Aqui a  ideia era tirar esses messages, não que seja um impeditivo... Mas ela tende a ser menos útil.

asouza commented on nosso-calendario/src/main/java/com/br/devefc/nossocalendario/model/User.java:
```diff
@@ -12,59 +12,62 @@
 	private String senha;
-	
+
 	@CreationTimestamp
```
Essa é uma solução esperta :). Por outro lado, o seu objeto nasce em um estado inconsistente. Será que precisa mesmo disso? Digo, fazer ``` dataCriacao = LocalDateTime.now()``` não resolveria?

asouza commented on nosso-calendario/src/main/java/com/br/devefc/nossocalendario/model/User.java:
```diff
@@ -12,59 +12,62 @@
+	public User(@NotBlank @NotEmpty @Email String login, @NotBlank @NotEmpty @Size(min = 6) String senha) {
 		this.login = login;
-		this.senha = senha;
+		this.senha = this.encodeSenha(senha);
```
Aqui, eu não quero ser chato hehe, mas preciso falar que o problema continua. Como que o lado de fora sabe que é para passar a senha em texto puro? E se ele passar encodada? Lembra que você precisa deixar uma pista para o lado de fora que essa não é uma string normal.

asouza commented on nosso-calendario/src/main/java/com/br/devefc/nossocalendario/model/User.java:
```diff
@@ -12,59 +12,62 @@
+		return encodedSenha;
 	}
 
 	public Long getId() {
```
**Idealmente, todo código escrito deveria ser chamado por alguém. Se não tem ninguém chamando, ele não deveria existir.**

Não precisamos ainda nem de getters e nem de setters. Não fazemos edição parcial do usuário e nem retornamos as informações dele para algum lugar.

asouza commented on nosso-calendario/src/main/java/com/br/devefc/nossocalendario/model/User.java:
```diff
@@ -12,59 +12,62 @@
+
 	public void setSenha(String senha) {
-		this.senha = senha;
+		this.senha = this.encodeSenha(senha);
```
aqui vale a mesma observação lá do construtor

asouza reviewed (changes requested):
Agora tamo quase lá :). A senha é o desafio!

Commit:
Remoção de messages Model do banco

Commit:
Correção do acesso_cadastro, para ficar com os dados consistentes

Commit:
Criação do helper para encoded de senha
 - Para que seja possivel informar ao lado de fora que a senha precisar
usar um hash para a senha

asouza commented on nosso-calendario/src/main/java/com/br/devefc/nossocalendario/model/User.java:
```diff
@@ -64,8 +59,8 @@ public String getSenha() {
 
-	public void setSenha(String senha) {
-		this.senha = this.encodeSenha(senha);
+	public void setSenha(EncodeSenha senha) {
```
não precisa do setter, você já recebe o EncodeSenha no construtor :).

asouza commented on nosso-calendario/src/main/java/com/br/devefc/nossocalendario/helpers/EncodeSenha.java:
```diff
@@ -0,0 +1,21 @@
+
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+
+public class EncodeSenha {
```
*Deixamos pistas que facilitem o uso do código onde não conseguimos resolver com compilação.*

Ficou bem legal. O nome podia ser um pouco mais semântico, mas já facilitou bastante!

arthurlima96 replied:
Como eu posso deixar mais semântico ?

asouza replied:
Acabei de subir o vídeo lá no canal :) => https://youtu.be/AivMq1vhK_Q

asouza reviewed (approved):
Boaaa!

Commit:
Remoção de sstter desnecessario
